modules/plugin/shapefile/src/test/java/org/geotools/data/shapefile/IndexManagerTest.java

```diff
@@ -0,0 +1,228 @@
+
+        mockDataStore = mock(ShapefileDataStore.class);
+        when(mockDataStore.getTypeName()).thenReturn(new NameImpl("statepop"));
+    }
+
+    @After
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        shpFiles.dispose();
+    }
+
+    @Test
+    public void testCreateSpatialIndex() throws Exception {
+        deleteQixFile();
+        IndexManager indexManager = new IndexManager(shpFiles, mockDataStore);
+
+        assertFalse(shpFiles.exists(QIX));
+        final boolean force = false;
+        assertTrue(indexManager.createSpatialIndex(force));
+        assertTrue(shpFiles.exists(QIX));
+        assertFalse(indexManager.createSpatialIndex(force));
+    }
+
+    @Test
+    public void testCreateSpatialIndex_force() throws Exception {
+        deleteQixFile();
+        IndexManager indexManager = new IndexManager(shpFiles, mockDataStore);
+
+        final boolean force = true;
+        assertTrue(indexManager.createSpatialIndex(force));
+
+        File file = null;
+        try {
+            file = shpFiles.acquireReadFile(QIX, indexManager.writer);
+            assertNotNull(file);
+        } finally {
+            shpFiles.unlockRead(file, indexManager.writer);
+        }
+
+        final long currentLastModified = file.lastModified();
+
+        for (int i = 0;
+                i < 10 && file.lastModified() == currentLastModified;
+                i++, Thread.sleep(100)) {
+            assertTrue(indexManager.createSpatialIndex(force));
+            assertTrue(shpFiles.exists(QIX));
+        }
+
+        assertTrue(file.lastModified() > currentLastModified);
+    }
+
+    @Test
+    public void testCreateSpatialIndex_stale_index_rebuilds() throws Exception {
+        IndexManager indexManager = new IndexManager(shpFiles, mockDataStore);
+        if (!shpFiles.exists(QIX)) {
+            assertTrue(indexManager.createSpatialIndex(true));
+        }
+
+        File shpFile = null;
+        File indexFile = null;
+        try {
+            shpFile = shpFiles.acquireReadFile(SHP, indexManager.writer);
+            indexFile = shpFiles.acquireReadFile(QIX, indexManager.writer);
+            assertTrue(indexFile.exists());
+        } finally {
+            shpFiles.unlockRead(indexFile, indexManager.writer);
+            shpFiles.unlockRead(shpFile, indexManager.writer);
+        }
+
+        // make the file stale (i.e. older than the .shp)
+        indexFile.setLastModified(shpFile.lastModified() - 100_000);
+        final long lastModified = indexFile.lastModified();
+
+        final boolean force = false;
+        assertTrue("stale index should force rebuild", indexManager.createSpatialIndex(force));
+        assertTrue(shpFiles.exists(QIX));
+
+        assertTrue(indexFile.lastModified() > lastModified);
+
+        assertFalse("index is not stale anymore", indexManager.createSpatialIndex(force));
+    }
+
+    @Test(timeout = 10_000) // timeout at 10s to fail in case of deadlock
```

Comment:
Same as above, allow longer timeouts... it should only time out if there is an issue, normally this test runs pretty quick, no?

Reply:
https://github.com/geotools/geotools/pull/3527/files#diff-cca0c8a31b2bd3b8cf2a2473aae2301caf3f80def759a5d36a2d82b3cc4e319bR137-R138